extensions/amp-sidebar/1.0/toolbar.js

```diff
@@ -0,0 +1,127 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {setStyles} from '../../../src/style';
+
+export class Toolbar {
+  /** @param {!Element} element */
+  constructor(element) {
+    this.element = element;
+
+    /** @private {?Element} */
+    this.sidebarElement_ = this.element.parentElement;
+
+    /** @private {string|undefined} */
+    this.toolbarMedia_ = undefined;
+
+    /** @private {?Element} */
+    this.toolbarClone_ = null;
+
+    /** @private {?Element} */
+    this.toolbarTarget_ = null;
+
+    /** @private {?Array} */
+    this.toolbarOnlyElements_ = null;
+
+    // Get our declared private variables
+    this.toolbarMedia_ = this.element.getAttribute('toolbar');
+
+    // Create a header element on the document for our toolbar
+    // TODO: Allow specifying a target for the toolbar
+    const fragment = this.sidebarElement_
+      .ownerDocument.createDocumentFragment();
+    this.toolbarTarget_ =
+      this.element.ownerDocument.createElement('header');
+    //Place the elements into the target
+    this.toolbarClone_ = this.element.cloneNode(true);
+    this.toolbarTarget_.appendChild(this.toolbarClone_);
+    if (!this.isToolbarShown_()) {
+      setStyles(this.toolbarTarget_, {
+        'display': 'none',
+      });
+    }
+
+    fragment.appendChild(this.toolbarTarget_);
+    this.sidebarElement_.parentElement
+        .insertBefore(fragment, this.sidebarElement_);
+
+    //Finally, find our tool-bar only elements
+    if (this.element.hasAttribute('toolbar-only')) {
+      this.toolbarOnlyElements_ = [];
+      this.toolbarOnlyElements_.push(this.element);
+    } else if (!this.element.hasAttribute('toolbar-only') &&
+      this.element.querySelectorAll('*[toolbar-only]').length > 0) {
+      this.toolbarOnlyElements_ = [];
+      // Check the nav's children for toolbar-only
+      Array.prototype.slice.call(this.element
+          .querySelectorAll('*[toolbar-only]'), 0)
+          .forEach(toolbarOnlyElement => {
+            this.toolbarOnlyElements_.push(toolbarOnlyElement);
+          });
+    }
+  }
+
+  /**
+   * Returns if the sidebar is currently in toolbar media query
+   * @returns {boolean}
+   * @private
+   */
+  isToolbarShown_() {
+    return this.element.ownerDocument.defaultView
+        .matchMedia(this.toolbarMedia_).matches;
+  }
+
+  /**
+   * Function called to check if we should show or hide the toolbar
+   * @param {!Function} onChangeCallback - function called if toolbar changes on check
+   */
+  checkToolbar(onChangeCallback) {
```

Comment:
Create 2 methods - attemptShow_ and hide_ , 
Open should check if the media matches and show the toolbar, hide should hide it. 

Whenever layout changes on the sidebar attempt to open the toolbar, it would be nice if we can achieve this via the observable class.  if not just loops are fine.